Output save file

arl opened the pull request:
- I created a checkable `QGroupBox` to contain the widgets for the serial output dump file settings. (there is still la bit of work on the layout i think)
- as said before, output internal buffer and output on screen are both cleared at session start whereas the dump file is not modified.
- implemented `<< operator` on `OutputMgr`
- dump file save/write implementation, i chose to do it simply in the `SessionManager`, where and when we receive the data from the serial port. `OutputManager` has nothing to do with file saving, it's about presenting the data. In the dump file there is no data conversion (at the exception of the end of lines). Also doing it in `OutputManager` would mean passing around the dump fiile name and format
- to not let the user lose the settings entered previously in the `ConnectDialog`, i am now keeping a pointer on the dialog and show/hide it when needed, in this way the settings remain. That was IMO easier than passing the `QHash` around between `MainWindow`, `ConnectDialog` and `SessionManager`.

arl committed:
add new widgets to connect dialog to handle dump

- ui : widgets added in a group box, but design needs review
	- filled with default values,read and added to emitted
	  configuration qHash

arl committed:
clear output buffer and text at session start

arl committed:
implement << operator on OutputManager

arl committed:
Remove OutputManager::handleNewData

arl committed:
implement dump file saving

arl committed:
current config is kept across ConnectDialog openings

In order to not alway have to re-set the same configuration settings,
ConnectDialog is now just shown and hidden :

	- to do this a pointer has to be kept inside the MainWindow
	- fixed : dump format ascii is not selectionned
	- some function renaming (settings to config to be homogenous)

arl committed:
refactor output saving into own function

arl committed:
fix incorrect number of line jumps in output

If a buffer ending with "\r\n" happens to be cut between '\r' and '\n'
the 2 resulting QString will be interpreted as having 2 line jumps,
instead of 1, when displayed in the output QTextEdit.
To avoid this, final '\r' are removed from the received buffer,
and prepended to the next one

lmancini commented:
Sul layout del `QGroupBox` non commento ma su Mac si presenta in modo veramente strano:

![screen shot 2015-06-26 at 17 01 45](https://cloud.githubusercontent.com/assets/829448/8380290/3a3b5a76-1c25-11e5-9b2e-302dfc12f4bf.png)

Conta che dovrebbe invece venir fuori una cosa del genere:

 ![screen shot 2015-06-26 at 17 04 25](https://cloud.githubusercontent.com/assets/829448/8380339/7f0d74a4-1c25-11e5-8587-9c44cfab410c.png)

Questo screenshot viene da un .ui rifatto da zero col Designer, ma in effetti provando ad aggiungere un `QGroupBox` al tuo .ui viene sempre con quell'aspetto piatto e il bordo viola.  Se vuoi lo guardiamo insieme.

lmancini commented on connectdialog.ui:
```diff
@@ -97,6 +100,116 @@
+      </size>
+     </property>
+     <property name="styleSheet">
+      <string notr="true">QGroupBox {  border: 1px solid rgb(130, 64, 166);}</string>
```
Penso sia questa riga il colpevole di quell'aspetto strano del `QGroupBox`...  È un'aggiunta volontaria?

arl replied:
Si questa riga di css l'avevo mesa per provare a vedere il frame del QGroupBox. Poi mi sono concentrato sul layout dei elementi e non sono arrivato al punto dove volevo arrivare. Quindi la riga l'ho lasciata perché volevo vedere con te come si aggiunge dal inizio e 'in a clean way' un groupbox e i suoi elementi...

arl replied:
sono ripartito da zero e ora è meglio

lmancini commented on outputmanager.cpp:
```diff
@@ -11,16 +11,59 @@
+    //           the 2 resulting QString will be interpreted as having 2 line jumps,
+    //           instead of 1, if displayed in a QTextEdit.
+    //           To avoid this, final '\r' are removed from the received buffer,
+    //           and prepended to the next one
```
Ok, questo in effetti spiega le stranezze sui caratteri di newline che vedevamo l'altra volta insieme.

lmancini commented:
Un piccolo dettaglio che ho notato adesso: quando scegli il radiobutton "txt" invece di "raw" per il formato di output, il nome del file di output resta `output.raw`; possiamo in alternativa cambiare l'estensione, oppure mettere un nome di default che non abbia un'estensione legata ad uno dei due formati di output (così va sempre bene).

arl committed:
tmp commit

arl committed:
connectdialog layout and more generic default dump file name

aurelien-rainone commented:
Ho cambiato tutte le cose viste con Lorenzo

gvaldambrini commented on connectdialog.cpp:
```diff
@@ -26,20 +26,26 @@ ConnectDialog::ConnectDialog(QWidget *parent) :
+    default_cfg["device"] = default_device;
+    default_cfg["baud_rate"] = QString::number(QSerialPort::Baud115200);
+    default_cfg["data_bits"] = QString::number(QSerialPort::Data8);
+    default_cfg["stop_bits"] = QString("1");
```
E' una piccolezza e me ne sono ricordato ora, ma in questi casi si può usare:
http://doc.qt.io/qt-5/qstring.html#QStringLiteral

arl replied:
fatto in a119cc1 e 48ee097

gvaldambrini commented on connectdialog.cpp:
```diff
@@ -94,14 +99,20 @@ void ConnectDialog::fillSettingsLists()
+    ui->dumpFile->setChecked(settings["dump_enabled"] == "1");
+    ui->dumpPath->setText(settings["dump_file"]);
+    ui->dumpRawFmt->setChecked(settings["dump_format"] == "raw");
+    ui->dumpTextFmt->setChecked(settings["dump_format"] == "ascii");
```
Perché non definisci il formato del dump come un enum e qua fai la conversione dell'intero a stringa con la QString::number? 
Come hai scritto te il codice, se per un typo scrivi "row" invece di "raw" la voce corrispondente non sarà mai selezionata e mai te ne accorgerai.

arl replied:
fatto in a119cc1 and 48ee097

arl committed:
use of QStringLiteral and enum for DumpFormat

arl committed:
forgot to replace enum in one line

gvaldambrini commented on outputmanager.cpp:
```diff
@@ -11,16 +11,59 @@
+
+    // notify that we have new data
+    emit dataConverted(to_send);
+}
```
C'è una ragione particolare per cui questo codice debba appartenere alla OutputManager? Ti faccio notare che se l'output manager contiene del codice che sa che successivamente (nella mainwindow) c'è un textedit si sta violando l'incapsulamento. Se c'è un problema con gli '\r' e il TextEdit mi viene da pensare che potrebbe risolverlo direttamente il MainWindow. 
In aggiunta, avevi provato a rimuovere del tutto gli '\r'?

arl replied:
Yes it's because i want to use the `OutputManager` as an helper class,
converting raw data into displayable data.
The manipulation I'm doing is necessary to present the data,  it happens to
take place in a `QTextEdit` and that's why I explained this in the comments
but i guess the code would remain the same if I had to present the data in
a terminal for example.

For the removal of `\r`,  I thought of it but if cutecom-ng had to display
the output of some program that use `\r` without `\n` (to print over some
already printed text for example) then the result would not be what is expected

gvaldambrini replied:
Mi torna quando dici che l'output manager converte "raw data" in dati visualizzabili, ma penso invece che il problema degli `\r` sia una caratteristica di Qt, che riscontri nel QTextEdit. Per esempio se stampi su terminale usando delle printf senza rimuovere gli `\r` mi aspetto che il terminale visualizzi le cose correttamente. 

Per quanto riguarda la rimozione dello `\r`, poiché fondamentalmente lo `\r` è mantenuto da moltissimi anni su Windows solo per compatibilità non credo sia un problema rimuoverlo (per esempio nella visualizzazione di file di test oggi tutti gli editor leggono correttamente un file che contiene solo `\n`). 
In ogni caso, mi sta bene lasciare inalterato quello che ricevi da terminale, ma se il problema degli `\r` è solo un problema di come Qt interpreta i newline (verifica stampando su terminale con i printf), spostiamo il codice nella main window. Ok?

arl replied:
Hai ragione quando dici che il problema dei newline è una caratteristica del QTextEdit, di Qt. Infatti, ho provato di rimuovere tutti `\r` è l'output si comporta normalmente.
Si comporta normalmente perché l'_expected output_ consista solo di linee stampate una dopo l'altra. Ma cosa succederà se cutecom dovesse stampare l'output di un program che usa `\r` per il suo _meaning_ originale (tornare a l'inizio della linea corrente senza saltare a la prossima), penso per esempio a un progress bar in cui il percentuale corrente si aggiorna

arl committed:
QTextEdit buffer manipulation moved to MainWindow

this is done to not break encapsulation: prevent OutputManager to be
tied only to QTextEdit

gvaldambrini commented:
@aurelien-rainone aggiungi sempre un commento quando vuoi che faccia la review e/o merge di una pull request, altrimenti non mi arrivano notifiche a riguardo!

aurelien-rainone commented:
ha ok va bene! ora lo farro
